plugin/nulls.go

```diff
@@ -6,73 +6,75 @@ import (
 	"github.com/apache/arrow/go/v13/arrow/memory"
 )
 
-// TODO(v4): use in v4
-//
-// nolint:unused
-func stripNullsFromLists(records []arrow.Record) {
-	for i := range records {
-		cols := records[i].Columns()
-		for c, col := range cols {
-			if col.DataType().ID() != arrow.LIST {
+func stripNullsFromLists(record arrow.Record) arrow.Record {
+	cols := record.Columns()
+	for c, col := range cols {
+		list, ok := col.(array.ListLike)
+		if !ok {
+			continue
+		}
+		if _, ok := list.(*array.Map); ok {
+			// maps also correspond to array.ListLike
+			continue
+		}
+
+		bldr := array.NewListBuilder(memory.DefaultAllocator, list.DataType().(arrow.ListLikeType).Elem())
+		for j := 0; j < list.Len(); j++ {
+			if list.IsNull(j) {
+				bldr.AppendNull()
 				continue
 			}
-
-			list := col.(*array.List)
-			bldr := array.NewListBuilder(memory.DefaultAllocator, list.DataType().(*arrow.ListType).Elem())
-			for j := 0; j < list.Len(); j++ {
-				if list.IsNull(j) {
-					bldr.AppendNull()
+			bldr.Append(true)
+			vBldr := bldr.ValueBuilder()
+			from, to := list.ValueOffsets(j)
+			slc := array.NewSlice(list.ListValues(), from, to)
+			for k := 0; k < int(to-from); k++ {
+				if slc.IsNull(k) {
 					continue
 				}
-				bldr.Append(true)
-				vBldr := bldr.ValueBuilder()
-				from, to := list.ValueOffsets(j)
-				slc := array.NewSlice(list.ListValues(), from, to)
-				for k := 0; k < int(to-from); k++ {
-					if slc.IsNull(k) {
-						continue
-					}
-					err := vBldr.AppendValueFromString(slc.ValueStr(k))
-					if err != nil {
-						panic(err)
-					}
+				err := vBldr.AppendValueFromString(slc.ValueStr(k))
+				if err != nil {
+					panic(err)
 				}
 			}
-			cols[c] = bldr.NewArray()
 		}
-		records[i] = array.NewRecord(records[i].Schema(), cols, records[i].NumRows())
+		cols[c] = bldr.NewArray()
 	}
+	return array.NewRecord(record.Schema(), cols, record.NumRows())
 }
 
 type AllowNullFunc func(arrow.DataType) bool
 
-// TODO(v4): use in v4
-//
-// nolint:unused
-func (f AllowNullFunc) replaceNullsByEmpty(records []arrow.Record) {
+func (f AllowNullFunc) replaceNullsByEmpty(record arrow.Record) arrow.Record {
```

Comment:
Is the method of a function necessary here?

Reply:
moved in 412d679eedeeefaa5a2c0308a6bbe230424abba7